### DISCUSSION TITLE
Use of reranker and Question about embeddings of DDL, SQL matched against Query

### DISCUSSION THREAD

**Discussion by sourabh-patil** [vanna-ai/vanna#410] [Q&A]
Hi,
Thanks for sharing this work. I had a question, which might be naive, I did not get how we get expected embeddings of DDL or SQL if we match it against natural language Query? Can anyone please shed some light on it? As per the base class of vanna, we first get the similar embeddings matched against stored (or here referred as trained) SQLs, DDLs and documentation.

**Comment by zainhoda** [comment]
Hi @sourabh-patil -- would you mind adding some additional context to your question? 

Have you seen this diagram?
![](https://github.com/vanna-ai/vanna/raw/main/img/vanna-readme-diagram.png)

**Reply by sourabh-patil** [reply]
Yes, I have gone through this diagram. The question was how we are getting expected  embeddings for DDL (which is not a natural language) that are similar to the Question (which is asked in natural language). From this diagram, in train phase we are storing DDL in vector db and for SQL we are storing it as question and SQL pair. At the time of Ask, we fetch embeddings similar to asked question from all three collections DDL, SQL and documentation. So, how do we make sure that DDLs which are stored (which are not in natural language) should match exactly with the question asked (which is in natural language). So do we assume that some words from question will match with some words in DDL? If yes then are there any better embedding models for this part of retreiver? 

In case of SQL, we are storing question and sql as pair so we will probably get back the expected SQL.

P.S. Should we use a reranker after we fetch the data from DB? I feel it should improve performance atleast in case of Documentation and SQL

**Comment by watertianyi** [comment]
@sourabh-patil 
Is training these ddl, sql and document files mainly for offline vector storage operations? Do these files need to be defined according to their own databases? Is each input query question converted into a vector and then the similarity calculated with the three file vectors in the vector library? Then combine it into prompt and send it to llm? Is this the workflow?

**Reply by sourabh-patil** [reply]
@watertianyi 
Training these ddl, document files is like a clue for LLM (through prompt) to understand the data and transform the natural language query to SQL query. 
Yes, these files should be defined as per database on which you are querying.
Yes, each new query does similarity search between three vector stores and gives you the most relevent ddl, sql qna pair and document files. As you said this is fed to LLM through prompt to make it understand the database and help it to write accurate db query.
